art_tree.go

```diff
@@ -17,40 +20,90 @@ func NewArtTree() *ArtTree {
 	return &ArtTree{root: nil, size: 0}
 }
 
+type Result struct {
+	Key   string
+	Value interface{}
+}
+
+func (t *ArtTree) EachChanResult() chan Result {
+	return t.EachChanResultFrom(t.root)
+}
+
+func (t *ArtTree) EachChanResultFrom(start *ArtNode) chan Result {
+	outChan := make(chan Result)
+	go func() {
+		if start != nil {
+			for n := range t.EachChanFrom(start) {
+				if n.IsLeaf() {
+					outChan <- Result{string(n.key), n.value}
+				}
+			}
+		}
+		close(outChan)
+	}()
+	return outChan
+}
```

Comment:
this converts from an object factory model to a transmit-receive.  the channel is open until it's closed, while the receiver waits for data and the returned channel is immediately usable.  even better, processing the results onto the channel are moved into the background.

this method is easier, especially if you're not going to build an array.  i'll point to the test case below.

if you were going to collect the results into an array, then this would be overkill.  which is why the single Search returns in the actual result.  but in a prefix search, we're expecting a series of matches, and then we'll pick over them or reassemble downstream.